include/gsl/gsl_assert

```diff
@@ -72,16 +72,49 @@ struct fail_fast : public std::runtime_error
 {
     explicit fail_fast(char const* const message) : std::runtime_error(message) {}
 };
-}
+namespace details
+{
+template <typename Exception>
+[[noreturn]] void throw_exception(Exception&& exception);
```

Comment:
As it is now, the build will fail at the linking stage if the user defines ```GSL_UNENFORCED_ON_CONTRACT_VIOLATION``` and uses ```narrow()```. My assumption was that if the user does not want "normal" enforcement with exceptions, they will have to provide their own implementation of ```throw_exception()```. What could be done instead is that we provide a specialization of ```throw_exception``` for ```gsl::narrowing_error``` when ```GSL_UNENFORCED_ON_CONTRACT_VIOLATION``` is defined.